test/e2e/framework/framework.go

```diff
@@ -169,33 +155,25 @@ func NewFramework(baseName string, options Options, client clientset.Interface)
 		Timeouts:                 NewTimeoutContextWithDefaults(),
 	}
 
-	f.AddAfterEach("dumpNamespaceInfo", func(f *Framework, failed bool) {
-		if !failed {
-			return
-		}
-		if !TestContext.DumpLogsOnFailure {
-			return
-		}
-		if !f.SkipNamespaceCreation {
-			for _, ns := range f.namespacesToDelete {
-				DumpAllNamespaceInfo(f.ClientSet, ns.Name)
-			}
-		}
-	})
-
 	ginkgo.BeforeEach(f.BeforeEach)
-	ginkgo.AfterEach(f.AfterEach)
```

Comment:
technical question, should 
```
	// In addition, AfterEach will not be called if a test never gets here.
	ginkgo.DeferCleanup(f.AfterEach)

	// Registered later and thus runs before deleting namespaces.
	ginkgo.DeferCleanup(f.dumpNamespaceInfo)
```

go here , in `NewDefaultFramework`  or inside `BeforeEach()`

The constructor seems the one configuring the hooks order, or do we want the dependency that f.AfterEach only happens if f.BeforeAch has called?

Reply:
It has to be in `BeforeEach` because `DeferCleanup` is not allowed outside of a `BeforeEach/AfterEach/It` node. That is a somewhat artificial limitation, but conceptually it makes sense: it is meant to be used in combination with some setup code if and only if that setup code actual ran.

That makes sense for us here. If `f.BeforeEach` never gets reached, for example because some earlier `BeforeEach` skipped the test, then there is no need to run `f.AfterEach`.